rs/nervous_system/instruction_stats/src/lib.rs

```diff
@@ -0,0 +1,130 @@
+//! Important limitation: This only only works within update methods, not query.
+//! This is because canisters (in particular, their heap) are not changed by
+//! queries. This is a fundamental platform limitation, not just a limitation in
+//! the implementation of this library.
+//!
+//! Basic usage:
+//!
+//! 1. At the beginning of a scope (e.g. the start of a canister method
+//!    implementation), add one line:
+//!    ```ignore
+//!    let _on_drop = UpdateInstructionStatsOnDrop::new(operation_name, additional_labels);
+//!    ```
+//!    (It is not weird for additional_labels to be BTreeMap::new().)
+//!
+//! 2. Call encode_instruction_metrics. This writes a data for a metric named
+//!    `candid_call_instructions`.
+
+use ic_metrics_encoder::MetricsEncoder;
+use ic_nervous_system_histogram::{Histogram, STANDARD_POSITIVE_BIN_INCLUSIVE_UPPER_BOUNDS};
+use lazy_static::lazy_static;
+use std::{cell::RefCell, collections::BTreeMap};
+
+#[cfg(test)]
+use crate::tests::call_context_instruction_counter;
+#[cfg(not(test))]
+use ic_cdk::api::call_context_instruction_counter;
```

Comment:
Does this use performance_counter(1)? Alternatively, are you sure that this function has the correct behavior for async functions?

Reply:
TBH, I find [these docs](https://docs.rs/ic-cdk/latest/ic_cdk/api/fn.call_context_instruction_counter.html) difficult to decipher, but as best as I can tell, yes.